### PR TITLE
Collective ops

### PR DESCRIPTION
**Pull request by typhoonzero**
Enable adding collective operators from python side. Support Allreduce for this PR.

Usage for multi-process or distributed training:

```python
opt = fluid.optimizer.Momentum(learning_rate=self.lr, momentum=0.9)
params_grads = opt.backward(avg_cost)
data_parallel_param_grads = []
for p, g in params_grads:
    # NOTE: scale will be done on loss scale in multi_devices_graph_pass using nranks.
    grad_reduce = fluid.layers.allreduce(g)
    data_parallel_param_grads.append([p, grad_reduce])
opt.apply_gradients(data_parallel_param_grads)
```

### CODE REVIEW

File: paddle/fluid/API.spec [hunk 1]
```diff
@@ -286,6 +286,7 @@ paddle.fluid.layers.StaticRNN.update_memory ArgSpec(args=['self', 'mem', 'var'],
 paddle.fluid.layers.reorder_lod_tensor_by_rank ArgSpec(args=['x', 'rank_table'], varargs=None, keywords=None, defaults=None)
 paddle.fluid.layers.Print ArgSpec(args=['input', 'first_n', 'message', 'summarize', 'print_tensor_name', 'print_tensor_type', 'print_tensor_shape', 'print_tensor_lod', 'print_phase'], varargs=None, keywords=None, defaults=(-1, None, -1, True, True, True, True, 'both'))
 paddle.fluid.layers.is_empty ArgSpec(args=['x', 'cond'], varargs=None, keywords='ignored', defaults=(None,))
+paddle.fluid.layers.allreduce ArgSpec(args=['x', 'out'], varargs=None, keywords=None, defaults=(None,))
```
Comment (panyx0718) [inline paddle/fluid/API.spec]: @velconia Can we use this for imperative?

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 2]
```diff
@@ -271,7 +271,15 @@ bool MultiDevSSAGraphBuilderBase::UseGPU() const {
   return use_gpu;
 }
 
-bool MultiDevSSAGraphBuilderBase::NeedCollectiveOps() const {
+bool MultiDevSSAGraphBuilderBase::NeedCollectiveOps(
+    std::vector<ir::Node *> ops) const {
+  for (auto *node : ops) {
+    if (node->Op()->Type() == "allreduce") {
+      // HACK: if graph have allreduce op, do not append automatic
+      // allreduce allreduce_op_handles
```
Comment (panyx0718) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc]: what's the long term plan?

File: paddle/fluid/operators/distributed_ops/allreduce_op.cc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  void apply() {
+    *data_ = tensor_->mutable_data<T>(place_);
+  }
+
+  void** data_;
+  framework::LoDTensor* tensor_;
+  platform::Place place_;
+};
+
+class AllReduceOp : public framework::OperatorBase {
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed_ops/allreduce_op.cc:43]: why not kernel?

File: paddle/fluid/framework/parallel_executor.cc [hunk 4]
```diff
@@ -254,6 +254,16 @@ ParallelExecutor::ParallelExecutor(
...
         build_strategy.trainer_id_));
+    // initialize device context's nccl comm, so that all operators
+    // can use it.
+    for (size_t dev_id = 0; dev_id < member_->places_.size(); ++dev_id) {
+      auto &nccl_ctx = member_->nccl_ctxs_->at(dev_id);
+      platform::DeviceContextPool &pool =
+          platform::DeviceContextPool::Instance();
+      auto *dev_ctx = static_cast<platform::CUDADeviceContext *>(
+          pool.Get(member_->places_[dev_id]));
+      dev_ctx->set_nccl_comm(nccl_ctx.comm());
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.cc]: is this the way to pass num places? can this be attr of all reduce op? can this be done in pass?

**Review by panyx0718 — COMMENTED** [review]
what's the reason for doing this?

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 2 above)
Comment (typhoonzero) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc]: > what's the reason for doing this?

To enable users to create arbitrary multi-device/distributed models at will, See the example above, currently the loss averaging is done by internal passes, if user want to average gradients other than loss, they may add it there, or people can run allreduce to any variable to achieve functions like "syncbn".

> what's the long term plan?

`multi_devices_graph_pass` only do replicate graph to multiple devices, python side `compiler.py:with_data_parallel` will add allreduce op for the program.

File: paddle/fluid/operators/distributed_ops/allreduce_op.cc (hunk 3 above)
Comment (typhoonzero) [reply paddle/fluid/operators/distributed_ops/allreduce_op.cc:43]: kernel should be better, since current implementation is simple, will move to use kernel when we have both CPU allreduce support.

File: paddle/fluid/framework/parallel_executor.cc (hunk 4 above)
Comment (typhoonzero) [reply paddle/fluid/framework/parallel_executor.cc]: `NCCLContext` is runtime, better not to put in attr of allreduce op or in passes. Is it properate to put `NCCLContext` in a singleton like `DeviceContextPool`?

Comment (Yancey0623) [inline paddle/fluid/framework/parallel_executor.cc]: Do this means computation and NCCL communication use the common CUDA stream? I think this may cause performance degradation.

File: python/paddle/fluid/tests/unittests/test_dist_base.py [hunk 5]
```diff
@@ -495,8 +534,8 @@ def _run_cluster_nccl2(self, model, envs, check_error_log):
         # print log
         sys.stderr.write('trainer 0 stderr: %s\n' % tr0_err)
         sys.stderr.write('trainer 1 stderr: %s\n' % tr1_err)
-        sys.stderr.write('trainer 0 stdout: %s\n' % tr0_out)
-        sys.stderr.write('trainer 1 stdout: %s\n' % tr1_out)
+        # sys.stderr.write('trainer 0 stdout: %s\n' % tr0_out)
```
Comment (Yancey0623) [inline python/paddle/fluid/tests/unittests/test_dist_base.py]: Please delete the unused code.

File: paddle/fluid/framework/parallel_executor.cc (hunk 4 above)
Comment (typhoonzero) [reply paddle/fluid/framework/parallel_executor.cc]: Will use the same stream for now. There's FIXME comment in `allreduce_op.cc` which will be fixed.

Comment (chengduoZH) [reply paddle/fluid/framework/parallel_executor.cc]: > Is it appropriate to put `NCCLContext` in a singleton like `DeviceContextPool`?

Maybe not appropriate, if the Python side has two ParallelExecutor, and they use different GPU cards, the NCCLContext should be different.

File: python/paddle/fluid/tests/unittests/test_dist_base.py (hunk 5 above)
Comment (typhoonzero) [reply python/paddle/fluid/tests/unittests/test_dist_base.py]: Done

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 6]
```diff
@@ -205,8 +205,9 @@ std::unique_ptr<ir::Graph> MultiDevSSAGraphBuilderBase::ApplyImpl(
             auto &p_name = backward_vars[i];
             auto &g_name = backward_vars[i + 1];
             VLOG(10) << "Bcast " << g_name << " for parameter " << p_name;
-
-            InsertCollectiveOp(&result, p_name, g_name);
+            if (NeedCollectiveOpHandle(g_name, sorted_ops)) {
```
Comment (Yancey0623) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc]: Merge `NeedCollectiveOpHandle ` and `NeedCollectiveOps ` ?

File: python/paddle/fluid/tests/unittests/dist_allreduce_op.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        opt = fluid.optimizer.Momentum(learning_rate=self.lr, momentum=0.9)
+        if single_device:
+            opt.minimize(avg_cost)
+        else:
+            # multi device or distributed multi device
+            params_grads = opt.backward(avg_cost)
+            data_parallel_param_grads = []
+            for p, g in params_grads:
+                # NOTE: scale will be done on loss scale in multi_devices_graph_pass using nranks.
+                grad_reduce = fluid.layers.control_flow._allreduce(g)
```
Comment (Yancey0623) [inline python/paddle/fluid/tests/unittests/dist_allreduce_op.py]: `layers.contro_flow` or `layers.io` ?

File: python/paddle/fluid/tests/unittests/test_dist_base.py [hunk 8]
```diff
@@ -427,29 +452,34 @@ def _run_cluster(self, model, envs, check_error_log):
             sys.stderr.write("ps1 stderr: %s\n" % fn.read())
 
         # print log
-        if stat0 == 0:
-            sys.stderr.write('trainer 0 stdout: %s\n' % pickle.loads(tr0_out))
+        # if stat0 == 0:
```
Comment (Yancey0623) [inline python/paddle/fluid/tests/unittests/test_dist_base.py]: Reminding the commended code ?

File: python/paddle/fluid/tests/unittests/test_dist_base.py [hunk 9]
```diff
@@ -87,7 +94,7 @@ def run_trainer(self, args):
                                     args.endpoints, args.trainers,
                                     args.sync_mode, args.dc_asgd)
             trainer_prog = t.get_trainer_program()
-        elif args.update_method == "nccl2":
+        elif args.update_method == "nccl2" or args.update_method == "nccl2_reduce_layer":
```
Comment (Yancey0623) [inline python/paddle/fluid/tests/unittests/test_dist_base.py:97]: maybe `nccl2_reduce_layer` mode should execute the `seq_allreduce_pass` to avoid the NCCL hang ?

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 6 above)
Comment (typhoonzero) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc]: Done.

File: python/paddle/fluid/tests/unittests/dist_allreduce_op.py (hunk 7 above)
Comment (typhoonzero) [reply python/paddle/fluid/tests/unittests/dist_allreduce_op.py]: moved to `collective.py`

File: python/paddle/fluid/tests/unittests/test_dist_base.py (hunk 9 above)
Comment (typhoonzero) [reply python/paddle/fluid/tests/unittests/test_dist_base.py:97]: It will be automatically done if nranks > 1

**Review by Yancey0623 — DISMISSED** [review]
LGTM, should find a better way to replace the confused `nccl2_reduce_layer` mode, we only need the `local`, `pserver` and `nccl2` mode.

**Review by Yancey0623 — APPROVED** [review]
LGTM.
